Design parser

gle-mini opened the pull request:
Parser for IRC message.

gle-mini committed:
Parser works [OK] check the norm [TODO] add header comment [TODO]

gle-mini committed:
parser [OK] norm [OK]

gle-mini committed:
testing parser [DOING]

gle-mini committed:
Merge branch 'main' into design-parser

gle-mini committed:
Manage merge conflicts [OK]

Saphirac commented on Makefile:
```diff
@@ -57,8 +57,8 @@ LIB_SRC = \
-				${addprefix IrcMessage/, \
-					IrcMessage \
+				${addprefix Message/, \
+					Message \
```
Please separate it as core.cpp and methods.cpp instead of Message.cpp and core.cpp

JonathanDUFOUR replied:
I agree with @Saphirac

Saphirac commented on private/Message.hpp:
```diff
@@ -15,23 +15,32 @@
+	// Accessors
+	std::string const        &get_prefix() const;
+	std::string const        &get_command() const;
+	std::vector<std::string> &get_params();
```
Why is get_params not const?

JonathanDUFOUR replied:
Good notie + it should also return a reference to constant vector.
Please change that.

Saphirac commented on private/Message.hpp:
```diff
@@ -15,23 +15,32 @@
+	// Methods
+	void add_param(std::string const &param);
+
+	void display() const;
```
Display being a debug function, it should be removed in the final main.
I do argue that we should keep it for the time being as we do need to test everything by the main fully merged with each of our parts

JonathanDUFOUR replied:
Actually, we should test everything that is related to what we are asking to merge onto the `main` branch BEFORE merging, and thus, every debug function should be removed before merging.

Saphirac commented on private/Message.hpp:
```diff
@@ -15,23 +15,32 @@
+	std::vector<std::string> &get_params();
+
+	// Setters
+	void set_prefix(std::string const &prefix);
```
Are those setter useful in the end?

JonathanDUFOUR replied:
I'm almost certain that they are useless, indeed.
Please remove them.

Saphirac commented on private/Message.hpp:
```diff
@@ -15,23 +15,32 @@
+	void set_params(std::vector<std::string> const &params);
+
+	// Methods
+	void add_param(std::string const &param);
```
If it's used only internally for parsing, it should be private, else I think it's useless.

JonathanDUFOUR replied:
It's true. Moreover, I would suggest to remove it if this method does what I think that it does: adding a string to the vector `this->_params`. If I'm right, then please, use directly the `std::vector` method to do that.

Saphirac commented on ft_irc_test:
We shouldn't forget to delete it before setting as finished as it could be considered cheating. Maybe you should put it in it's own branches and we should not push it in the main ?

JonathanDUFOUR replied:
I think putting it on the `main` branch isn't a bad thing. We just have to delete it when we are sure that every test is passed successfully.

Saphirac commented on src/lib/class/Message/Message.cpp:
```diff
@@ -0,0 +1,57 @@
+ */
+Message::Message(std::string const &msg)
+{
+	size_t pos = 0;
```
I would've called a function that does the parser instead of doing it all in the constructor

JonathanDUFOUR replied:
It is unnecessary according to me: the parsing is well segmented here, and the one-line comments below make the different segments easy to spot. Moreover, the whole parsing takes less than 50 lines.

Saphirac commented on src/lib/class/Message/core.cpp:
```diff
@@ -0,0 +1,82 @@
+ */
+void Message::add_param(const std::string &param) { _params.push_back(param); }
+
+#include <iostream>
```
include should be at the top of the file.

JonathanDUFOUR replied:
In fact, in this case, it was for a debug function, and thus, it is understandable to put debug `#include` at the top of the first function/method that needs it.

JonathanDUFOUR replied:
But as the `display` method should be removed, this `#include` should be removed too.

Saphirac commented on private/Message.hpp:
Message.hpp should be in the class directory inside the include dir

JonathanDUFOUR replied:
I agree that it should be in a `class` directory, but not in the `include` directory. Instead, it should be in `private/class`.

Saphirac commented on src/lib/class/Message/core.cpp:
```diff
@@ -0,0 +1,82 @@
+
+void Message::display() const
+{
+	std::cout << "Prefix: " << this->_prefix << std::endl;
```
<< std::endl are a bit too much if there is no error that needs to be signaled immediately, you should prefer a simple \n

Saphirac reviewed (changes requested):
Please check my comments and either answer or change accordingly. Jonathan will also check before merging approval

gle-mini committed:
Merge branch 'main' into design-parser

JonathanDUFOUR commented on src/lib/class/Message/Message.cpp:
```diff
@@ -0,0 +1,57 @@
+	{
+		spacePos = msg.find(' ', pos);
+		this->set_prefix(msg.substr(pos + 1, spacePos - pos - 1));
+		pos = spacePos + 1;
```
There is a unhandled possible overflow here.
Overflows cause undefined behavior by definition.
Please, handle it.

JonathanDUFOUR commented on src/lib/class/Message/Message.cpp:
```diff
@@ -0,0 +1,57 @@
+ */
+Message::Message(std::string const &msg)
+{
+	size_t pos = 0;
```
`pos` is actually not a self-explicit variable name. Of course it stands for `position`, that's not the problem, but the position of what? For example, your other variable `spacePos` is quite self-explicit (it even could have been named `next_space_pos` for more accuracy).

JonathanDUFOUR commented on src/lib/class/Message/Message.cpp:
```diff
@@ -0,0 +1,57 @@
+Message::Message(std::string const &msg)
+{
+	size_t pos = 0;
+	size_t spacePos = 0;
```
Norm error + initializing it now to zero is irrelevant as you re-assign it just after, whatever happens, with `msg.find(' ', pos)`.

JonathanDUFOUR commented on src/lib/class/Message/Message.cpp:
```diff
@@ -0,0 +1,57 @@
+	{
+		this->set_command(msg.substr(pos));
+		return;
+	}
```
You better should check the opposite: check that `spacePos` is equal to `std::string::npos`. Doing so, you will return un the `if` body, and thus, you won't even need to have an explicit `else`.

JonathanDUFOUR commented on src/lib/class/Message/Message.cpp:
```diff
@@ -0,0 +1,57 @@
+	if (msg[pos] == ':')
+	{
+		spacePos = msg.find(' ', pos);
+		this->set_prefix(msg.substr(pos + 1, spacePos - pos - 1));
```
You'd better increment `pos` instead of doing both a `+ 1` and a `- 1`.

JonathanDUFOUR commented on src/lib/class/Message/Message.cpp:
```diff
@@ -0,0 +1,57 @@
+		if (msg[pos] == ':')
+		{
+			this->add_param(msg.substr(pos + 1));
+			break;
```
You'd better `return` instead of `break` at this point, because if you fall in this `if`, that means that you have reached the last argument (starting with a `:`). Doing so will also allows you to remove your following explicit `else`.

JonathanDUFOUR commented on src/lib/class/Message/Message.cpp:
Please remove this file, and put the constructor in the `core.cpp` file.

JonathanDUFOUR commented on src/lib/class/Message/core.cpp:
```diff
@@ -0,0 +1,82 @@
+
+/**
+ * Destructor for the Message class.
+ */
```
Doxygen comments are unnecessary for the core methods.

JonathanDUFOUR commented on src/lib/class/Message/core.cpp:
```diff
@@ -0,0 +1,82 @@
+/* ************************************************************************** */
+
+#include "Message.hpp"
+
```
Please, add section comments like the following example:
```c++

// Constructors //

Sample::Sample(void) {}

// Destructor //

Sample::~Sample(void) {}

// Accessors //

int Sample::get_a(void) const { return this->_a; }
int Sample::get_b(void) const { return this->_b; }
int Sample::get_c(void) const { return this->_c; }

// Mutators //

void Sample::set_a(int const a) { this->_a = a; }
void Sample::set_b(int const b) { this->_b = b; }
void Sample::set_c(int const c) { this->_c = c; }

// Operators //

Sample &Sample::operator=(Sample const &rhs)
{
  this->_a = rhs->_a;
  this->_b = rhs->_b;

  return *this;
}
```

JonathanDUFOUR commented on src/lib/class/Message/core.cpp:
```diff
@@ -0,0 +1,82 @@
+ *
+ * @param msg A constant reference to a std::string containing the message to be parsed.
+ */
+std::string const &Message::get_prefix() const { return _prefix; }
```
Please, add the `this->` prefix everytime you access a field, or call a method from the class itself.

JonathanDUFOUR commented on src/lib/class/Message/core.cpp:
```diff
@@ -0,0 +1,82 @@
+ *
+ * @return A constant reference to a std::string containing the command of the message.
+ */
+std::string const &Message::get_command() const { return _command; }
```
Please, add the this-> prefix everytime you access a field, or call a method from the class itself.

JonathanDUFOUR commented on src/lib/class/Message/core.cpp:
```diff
@@ -0,0 +1,82 @@
+ *
+ * @return A reference to a vector of std::string containing the parameters of the message.
+ */
+std::vector<std::string> &Message::get_params() { return _params; }
```
Please, add the this-> prefix everytime you access a field, or call a method from the class itself.

JonathanDUFOUR commented on src/lib/class/Message/core.cpp:
```diff
@@ -0,0 +1,82 @@
+ *
+ * @param prefix A constant reference to a std::string containing the prefix to be set.
+ */
+void Message::set_prefix(const std::string &prefix) { _prefix = prefix; }
```
Please, add the this-> prefix everytime you access a field, or call a method from the class itself.

JonathanDUFOUR commented on src/lib/class/Message/core.cpp:
```diff
@@ -0,0 +1,82 @@
+ *
+ * @param command A constant reference to a std::string containing the command to be set.
+ */
+void Message::set_command(const std::string &command) { _command = command; }
```
Please, add the this-> prefix everytime you access a field, or call a method from the class itself.

JonathanDUFOUR commented on src/lib/class/Message/core.cpp:
```diff
@@ -0,0 +1,82 @@
+ *
+ * @param params A constant reference to a vector of std::string containing the parameters to be set.
+ */
+void Message::set_params(const std::vector<std::string> &params) { _params = params; }
```
Please, add the this-> prefix everytime you access a field, or call a method from the class itself.

JonathanDUFOUR commented on src/lib/class/Message/core.cpp:
```diff
@@ -0,0 +1,82 @@
+ *
+ * @param param A constant reference to a std::string containing the parameter to be added.
+ */
+void Message::add_param(const std::string &param) { _params.push_back(param); }
```
Please, add the this-> prefix everytime you access a field, or call a method from the class itself.

JonathanDUFOUR commented on src/main.cpp:
No any change except the 42 header?? Please, in the future, check your diffs before committing, and whenever you see that you your only change on a file is the 42 header, please `git restore` it.

JonathanDUFOUR reviewed (changes requested):
Gloabally, it's a good work. Some mistakes remain, but it should be fast to fix ;) Keep it up!

gle-mini committed:
change for pull request [OK]

gle-mini committed:
Merge branch 'main' into design-parser

gle-mini committed:
Prefix class added and tested [OK]

Saphirac committed:
Corrected for PR [OK]

Saphirac reviewed (commented):
All good !

JonathanDUFOUR commented on Makefile:
```diff
@@ -199,10 +202,15 @@ LIB_SRC := \
 		} \
-	}
+	} \
+
```
The last brace must not have a trailing backslash. Please remove it, and remove this additional empty line.

JonathanDUFOUR commented on private/class/Message.hpp:
```diff
@@ -19,6 +19,7 @@ class Message
 {
 public:
 	// Constructors
+
```
Please remove this empty line.

JonathanDUFOUR commented on src/lib/class/Message/core.cpp:
```diff
@@ -15,51 +15,64 @@
+	{
+		this->_command = msg.substr(pos);
 		return;
 	}
```
You didn't change this `if-else` statement as I asked to.

JonathanDUFOUR commented on src/lib/class/Message/core.cpp:
```diff
@@ -15,51 +15,64 @@
+		{
+			this->_parameters.push_back(msg.substr(pos, space_pos - pos));
+			pos = space_pos + 1;
 		}
```
No need for an `else` statement, because you `return` if you fall in the previous `if` statement. Please remove it.

JonathanDUFOUR commented on src/lib/class/Message/core.cpp:
```diff
@@ -15,51 +15,64 @@
 		{
-			this->_parameters.push_back(raw_msg.substr(pos + 1));
-			break;
+			this->_parameters.push_back(msg.substr(pos));
```
You must `msg.substr(pos + 1)` to skip the leading colon, not just `msg.substr(pos)`.

JonathanDUFOUR commented on src/lib/class/Message/core.cpp:
```diff
@@ -15,51 +15,64 @@
+// Destructors //
 
-Message::~Message(void) {}
+Message::~Message() {}
```
Add the `void` keyword in the parameters to be explicit.

JonathanDUFOUR commented on src/lib/class/Server/methods.cpp:
```diff
@@ -176,6 +176,7 @@ void Server::_compute_next_msg_for_a_client(Client &client)
 	if (!raw_msg.empty())
 	{
 		Message const         msg(raw_msg);
+		
```
Please remove this additional empty line.

JonathanDUFOUR commented on src/lib/class/specialized_string/Prefix/methods.cpp:
```diff
@@ -0,0 +1,128 @@
+static bool is_alphanumeric(char c)
+{
+	return (c >= '0' && c <= '9') || (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z');
+}
```
The `isalnum` standard function already exists for doing that. Please use it instead of your homemade function.

JonathanDUFOUR commented on src/main.cpp:
No changes, please revert.

JonathanDUFOUR commented on src/lib/class/Server/methods.cpp:
No changes, please revert.

JonathanDUFOUR committed:
Make some requested changes from the merge request #21 [OK]
Fix the `Prefix::is_valid` method                      [TODO]

JonathanDUFOUR reviewed (changes requested):
Please, complete the `Prefix::is_valid` method before merging to the `main` branch.

gle-mini committed:
fix the HostName is_valid function

gle-mini committed:
is_valid return false if the nickname is empty

gle-mini committed:
add dash in abnf_components

gle-mini committed:
delete the '- 1' in the substr to add ':' in the prefix

gle-mini committed:
change '-' to 'dash

gle-mini committed:
accept dash in nickname

gle-mini committed:
change is_valid function to manage empty string and add all the nolfcrspat characters

gle-mini committed:
is_valid prefix function [OK]

Saphirac commented on src/lib/class/specialized_string/Prefix/methods.cpp:
```diff
@@ -0,0 +1,57 @@
+	if (!NickName(server_or_nick).is_valid())
+		return false;
+
+	std::string remainder = this->substr(end_server_or_nick + 1);
```
I think you mean remaining

Saphirac commented on src/lib/class/specialized_string/Prefix/methods.cpp:
```diff
@@ -0,0 +1,57 @@
+	if (this->empty())
+		return true;
+
+	if (this->at(0) != ':')
```
You could group both if in one

JonathanDUFOUR replied:
More than this, the first `if` is useless.

Saphirac commented on src/lib/class/specialized_string/Prefix/methods.cpp:
```diff
@@ -0,0 +1,57 @@
+#include "class/specialized_string/NickName.hpp"
+#include "class/specialized_string/Prefix.hpp"
+#include "class/specialized_string/UserName.hpp"
+#include <cctype>
```
unused include

Saphirac reviewed (changes requested):
Please change the few things i asked

gle-mini committed:
change remainder to remaining

gle-mini committed:
Fix `NickName::is_valid` method                               [OK]
Revert `UserName::is_valid` method                            [OK]
Remove unnecessary variables in the `Prefix::is_valid` method [OK]
Add `const` qualifiers when possible                          [OK]

gle-mini committed:
delete unused library

Saphirac reviewed (dismissed):
Seems good to me :)

JonathanDUFOUR commented on src/lib/class/specialized_string/Prefix/methods.cpp:
```diff
@@ -0,0 +1,56 @@
+	if (!NickName(server_or_nick).is_valid())
+		return false;
+
+	std::string remaining = this->substr(end_server_or_nick + 1);
```
Missing `const`

JonathanDUFOUR commented on src/lib/class/specialized_string/NickName/methods.cpp:
```diff
@@ -21,6 +20,8 @@
 bool NickName::is_valid(void) const
 {
+	if (this->empty())
+		return false;
```
Useless `if`, please remove it.

gle-mini replied:
Yes it is because the `std::string::at` method throws an exception if called with a position greater or equal to the length.

JonathanDUFOUR commented on src/lib/class/specialized_string/NickName/methods.cpp:
```diff
@@ -21,6 +20,8 @@
+		return false;
 	return this->size() <= MAXIMUM_LENGTH_FOR_NICKNAME && (letter + special).find(this->at(0)) != this->npos
-	    && this->find_first_not_of(letter + digit + special, 1) == this->npos;
+	    && this->find_first_not_of(letter + digit + special + dash, 1) == this->npos;
```
Incorrect, nicknames may contain dash characters if it isn't the first character.

gle-mini replied:
Yes, that's exactly why I added the `dash` to the allowed characters here :)

JonathanDUFOUR replied:
Indeed, you are right, my bad, we must check that the string isn't empty before calling the `std::string::at` method with the position `0`.

JonathanDUFOUR commented on src/lib/class/specialized_string/Prefix/methods.cpp:
```diff
@@ -0,0 +1,56 @@
+		return false;
+
+	size_t      end_server_or_nick = this->find_first_of("!@", 1);
+	std::string server_or_nick = this->substr(1, end_server_or_nick - 1);
```
Missing `const`

JonathanDUFOUR commented on src/lib/class/specialized_string/Prefix/methods.cpp:
```diff
@@ -0,0 +1,56 @@
+			return false;
+
+		std::string username = remaining.substr(0, at_pos);
+		std::string hostname = remaining.substr(at_pos + 1);
```
Missing `const` + unnecessary to store the result of `substr` here.
You'd better do
```c++
return UserName(remaining.substr(0, at_pos)).is_valid() && HostName(remaining.substr(at_pos + 1)).is_valid;
```

JonathanDUFOUR commented on src/lib/class/specialized_string/UserName/methods.cpp:
```diff
@@ -17,4 +17,16 @@
+			return false;
+	}
+	return true;
+}
```
Incorrect! Please change to
```c++
return !this->empty() && this->find('@') == std::string::npos;
```

JonathanDUFOUR reviewed (changes requested):
Go back to work!

gle-mini committed:
Merge branch 'design-parser' of github.com:Saphirac/ft_irc into design-parser

gle-mini committed:
Merge branch 'main' into design-parser

gle-mini committed:
Add const in Prefix::is_valid function

Saphirac reviewed (dismissed):
Good!

gle-mini committed:
remove empty line in _compute_next_msg_for_a_client [OK]

gle-mini committed:
PR [OK]

JonathanDUFOUR reviewed (approved):
Everything is good! Great work!
